executor/executor_coordinator.go

```diff
@@ -141,33 +144,57 @@ func (ec *Coordinator) run(ctx context.Context) {
 				// get message delay from leader elector
 				readyTimestamp := ec.leaderElector.GetReadyTimestamp(id, time.Now().Unix())
 
-				heap.Push(ec.delayedMessageHeap, &message_heap.MessageWithTimestamp{
-					Payload:   &msg,
-					ReadyTime: readyTimestamp,
+				heap.Push(ec.delayedMessageHeap, &message_heap.MessageWithTimestamps{
+					Message:       &msg,
+					ReadyTime:     readyTimestamp,
+					ExpiryTime:    readyTimestamp + int64(ec.expiryDuration.Seconds()),
+					RetryInterval: ec.leaderElector.GetRetryDelay(msg.DestChainSelector),
+					MessageID:     &id,
 				})
 			}
 		case <-ticker.C:
 			// todo: get this current time from a single source across all executors
 			currentTime := time.Now().Unix()
 			readyMessages := ec.delayedMessageHeap.PopAllReady(currentTime)
-			for _, message := range readyMessages {
+			for _, payload := range readyMessages {
+				if currentTime > payload.ExpiryTime {
+					ec.lggr.Infow("message has expired", "messageID", *payload.MessageID)
+					continue
+				}
+				// TODO: use a worker pool here to avoid unbounded memory allocation
 				go func() {
-					message := message
-					id, _ := message.MessageID() // can we make this less bad?
+					message, currentTime, id := *payload.Message, currentTime, *payload.MessageID
+
 					ec.lggr.Infow("processing message with ID", "messageID", id)
-					err := ec.executor.AttemptExecuteMessage(ctx, message)
-					if errors.Is(err, ErrMsgAlreadyExecuted) {
-						ec.lggr.Infow("message already executed, skipping", "messageID", id)
-						return
-					} else if errors.Is(err, ErrInsufficientVerifiers) {
-						ec.lggr.Infow("not enough verifiers to execute message, will wait until next notification", "messageID", id, "error", err)
-						return
-					} else if err != nil {
-						ec.lggr.Errorw("failed to process message", "messageID", id, "error", err)
-						ec.monitoring.Metrics().IncrementMessagesProcessingFailed(ctx)
-						return
+					shouldRetry, shouldExecute, err := ec.executor.GetMessageStatus(ctx, message, currentTime)
+					if err != nil {
+						ec.lggr.Errorw("failed to check message status", "messageID", id, "error", err)
+					}
+
+					if shouldRetry {
+						retryTime := currentTime + payload.RetryInterval
```

Comment:
Do we want to keep retries simple like this (i.e. just adding a fixed configurable delay) vs. doing something like exponential backoff or more configurability?